File: packages/react-native-transparent-video-player/android/src/main/java/expo/modules/transparentvideo/TransparentVideoModule.kt
package expo.modules.transparentvideo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class TransparentVideoModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TransparentVideo")

    View(TransparentVideoView::class) {
      Events("onVideoEnd", "onFirstFrame", "onError")

      Prop("sourceUri") { view: TransparentVideoView, uri: String? ->
        view.setSource(uri)
      }

      Prop("loop") { view: TransparentVideoView, loop: Boolean ->
        view.loop = loop
      }

      Prop("paused") { view: TransparentVideoView, paused: Boolean ->
        view.paused = paused
      }

      // Only STAGES the nonce; the replay side effect runs in commitProps()
      // (OnViewDidUpdateProps) so it is always evaluated after sourceUri
      // regardless of the order props are applied within a batch — a replay
      // must never seek an outgoing clip.
      Prop("replayNonce") { view: TransparentVideoView, nonce: Int? ->
        view.pendingReplayNonce = nonce
      }

      OnViewDidUpdateProps { view: TransparentVideoView ->
        view.commitProps()
      }

      OnViewDestroys { view: TransparentVideoView ->
        view.release()
      }
    }
  }
}

File: packages/react-native-transparent-video-player/ios/TransparentVideoModule.swift
import ExpoModulesCore

public class TransparentVideoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TransparentVideo")

    View(TransparentVideoView.self) {
      // onError currently fires only on Android (decoder failures, with
      // bounded auto-retry); registered on iOS so the prop is valid there.
      Events("onVideoEnd", "onFirstFrame", "onError")

      // Prop setters only STAGE values; side effects run in commitProps()
      // (OnViewDidUpdateProps) so replayNonce is always evaluated after
      // sourceUri regardless of the order props are applied within a batch.
      Prop("sourceUri") { (view: TransparentVideoView, uri: String?) in
        view.pendingSourceUri = uri
      }

      Prop("loop") { (view: TransparentVideoView, loop: Bool) in
        view.loop = loop
      }

      Prop("paused") { (view: TransparentVideoView, paused: Bool) in
        view.paused = paused
      }

      Prop("replayNonce") { (view: TransparentVideoView, nonce: Int?) in
        view.pendingReplayNonce = nonce
      }

      OnViewDidUpdateProps { (view: TransparentVideoView) in
        view.commitProps()
      }
    }
  }
}
